### ISSUE TITLE
Compiler incorrectly optimizes [val].join("")

### ISSUE THREAD

**Issue by nbeloglazov** [google/closure-compiler#1676]
Hi

I believe compiler incorrectly handles following optimization: `[val].join('')` to `val + ''`. This conversion doesn't work when val has different `toString()` and `valueOf()` methods. For `join` browser uses `val.toString()` while for `+` it uses `val.valueOf()`. 

Example:

``` js
var val = {
  toString: function() { return 'foo'; },
  valueOf: function() { return 42; }
};

function myJoin(v) {
  return v['join']('');
}

console.log([val].join(''));
console.log(myJoin([val]));
```

When compiling with simpl optimizations and running in console I get the following:

``` text
42
foo
```

Correct output is: 

``` text
foo
foo
```

[UI service](http://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F%252F%2520ADD%2520YOUR%2520CODE%2520HERE%250Avar%2520val%2520%253D%2520%257B%250A%2520%2520toString%253A%2520function%28%29%2520%257B%2520return%2520'foo'%253B%257D%252C%250A%2520%2520valueOf%253A%2520function%28%29%2520%257B%2520return%252042%253B%2520%257D%250A%257D%253B%250Aconsole.log%28%255Bval%255D.join%28%2522%2522%29%29%253B%250A%250Afunction%2520myJoin%28v%29%2520%257B%250A%2520%2520return%2520v%255B'join'%255D%28%2522%2522%29%253B%250A%257D%250Aconsole.log%28myJoin%28%255Bval%255D%29%29%253B%250A%250A)

Example is hacky as I didn't find better way to prevent compiler from optimizing second `join` call.

**Comment by concavelenz** [comment]
Thanks for the report.

**Comment by lauraharker** [comment]
Looks like the same issue as #2782, which was closed by 736de5a.
